packages/cursorless-engine/src/processTargets/modifiers/scopeHandlers/BoundedNonWhitespaceSequenceScopeHandler.ts

```diff
@@ -0,0 +1,178 @@
+import {
+  Direction,
+  Position,
+  ScopeType,
+  TextEditor,
+  type Range,
+} from "@cursorless/common";
+import { ParagraphTarget, TokenTarget } from "../../targets";
+import { BaseScopeHandler } from "./BaseScopeHandler";
+import { compareTargetScopes } from "./compareTargetScopes";
+import type { TargetScope } from "./scope.types";
+import type {
+  ScopeHandler,
+  ScopeIteratorRequirements,
+} from "./scopeHandler.types";
+import type { ScopeHandlerFactory } from "./ScopeHandlerFactory";
+
+abstract class BoundedBaseScopeHandler extends BaseScopeHandler {
+  protected readonly isHierarchical = true;
+  private readonly targetScopeHandler: ScopeHandler;
+  private readonly surroundingPairInteriorScopeHandler: ScopeHandler;
+
+  constructor(
+    private scopeHandlerFactory: ScopeHandlerFactory,
+    private languageId: string,
+    private targetScopeType: ScopeType,
+  ) {
+    super();
+
+    this.targetScopeHandler = this.scopeHandlerFactory.create(
+      this.targetScopeType,
+      this.languageId,
+    )!;
+    this.surroundingPairInteriorScopeHandler = this.scopeHandlerFactory.create(
+      {
+        type: "surroundingPairInterior",
+        delimiter: "any",
+      },
+      this.languageId,
+    )!;
+  }
+
+  get iterationScopeType(): ScopeType {
+    if (this.targetScopeHandler.iterationScopeType.type === "custom") {
+      throw Error(
+        "Iteration scope type can't be custom for BoundedBaseScopeHandler",
+      );
+    }
+    return {
+      type: "oneOf",
+      scopeTypes: [
+        this.targetScopeHandler.iterationScopeType,
+        {
+          type: "surroundingPairInterior",
+          delimiter: "any",
+        },
+      ],
+    };
+  }
+
+  *generateScopeCandidates(
+    editor: TextEditor,
+    position: Position,
+    direction: Direction,
+    hints: ScopeIteratorRequirements,
+  ): Iterable<TargetScope> {
+    const targetScopes = this.targetScopeHandler.generateScopes(
+      editor,
+      position,
+      direction,
+      {
+        ...hints,
+        // Don't skip containing paint since it might have non contained nested scopes.
+        containment:
+          hints.containment !== "disallowed" ? hints.containment : undefined,
+      },
+    );
+
+    const pairScopes = Array.from(
+      this.surroundingPairInteriorScopeHandler.generateScopes(
+        editor,
+        position,
+        direction,
+        {
+          ...hints,
+          // For every (skipAncestorScopes=true) we don't want to go outside of the surrounding pair
+          containment:
+            hints.containment == null && hints.skipAncestorScopes
+              ? "required"
+              : hints.containment,
```

Comment:
what happens if you leave this out? A bit tough to reason about what's happening here 😅

Reply:
`every short block` takes outside of the containing pair